README.md
# remote_serial

[![License](./apache20.svg)](./LICENSE.txt)

This is an ultimate implementation of serial line interface for ROS2.

It provides ROS2 interfaces for inter-process access, introspection and
debugging.
It performs wisely in case of serial line saturation in any of
the I/O directions, minimizing data loses and blocking behavior,
ensuring optimum performance.

This package implements a generic purpose serial port driver using
character device files.
However this package also implements most of the logic required for other
serial line drivers (see the "Examples" section below).

## ROS2 interfaces

### Parameters

All classes:

- `serial_prefix`: the prefix to ROS2 interfaces exposed by this driver
  ("/serial/&lt;node-name&gt;" by default)

The factory class:

- `serial_is_remote`: instructs whether
  to instantiate the driver locally or to connect to a remote instance
  ("true" by default)

Serial port driver:

- `serial_dev_name`: the character device path ("/dev/ttyS0" by default)
- serial port settings:
  - `serial_skip_init`: skip the port initialization (ignore the below parameters)
  - `serial_baud_rate`: baud rate ("115200" by default)
  - `serial_data`: data bits ("8" by default)
  - `serial_parity`: parity bit ("false" by default)
  - `serial_stop`: stop bits ("1" by default)
  - `serial_flow_control`: hardware flow control ("false" by default)
  - `serial_sw_flow_control`: software flow control ("false" by default)
  - `serial_bs`: the size of read buffer ("1024" by default)

### Topics

#### Subscribers

- `<namespace>/<serial_prefix>/inject/input`: injects the given bytes into
  the serial line as if it was written to the line by the driver
- `<namespace>/<serial_prefix>/inject/output`: injects the given bytes into
  the serial line as if it was received by the driver from the line

#### Publishers

- `<namespace>/<serial_prefix>/inspect/input`: publishes data
  received by the driver from the line
- `<namespace>/<serial_prefix>/inspect/output`: publishes data
  sent by the driver to the line

## Basic setup

Launch it as a separate node for each serial port:

```
$ ros2 run remote_serial remote_serial_standalone
```

or

```
$ ros2 run remote_serial remote_serial_standalone \
  --ros-args \
  --remap serial:__node:=serial_com1 \
  -p serial_is_remote:=false \
  -p serial_prefix:=/serial/com1 \
  -p serial_dev_name:=/dev/ttyS0 \
  -p serial_baud_rate:=115200 \
  -p serial_data:=8 \
  -p serial_parity:=false \
  -p serial_stop:=1 \
  -p serial_flow_control:=true
```

```mermaid
flowchart TB
    cli["<p style='text-align:left'>$ ros2 topic echo /serial/com1/inspect/input\n$ ros2 topic echo /serial/com1/inspect/output\n$ ros2 topic pub /serial/com1/inject/output \\n&nbsp;&nbsp;&nbsp;std_msgs/msg/UInt8MultiArray {'data':'ATH+CHUP\r\ n'}</p>"] -. "DDS" .-> topic_serial[/ROS2 interfaces:\n/serial/com1/.../]
    app["Your process"] -- "DDS\n(with context switch)" --> topic_serial
    subgraph serial["Process: remote_serial_standalone"]
      topic_serial --> driver["Serial port driver"]
    end
    driver --> file{{"Character device: /dev/ttyS0"}}
```

## Advanced setup

The more advanced setup is to initialize it as a separate node in the very executable which will be communicating with the port all the time
(e.g. a MODBUS RTU implementation).

This setup allows DDS to forward the messages between nodes
without context switches should your DDS implementation support that.
See an example of such a setup in the Modbus RTU package:

```mermaid
flowchart TB
    cli_serial["<p style='text-align:left'># Serial debugging and troubleshooting\n$ ros2 topic echo /serial/inspect/input\n...</p>"] -. "DDS" ..-> topic_serial[/ROS2 interfaces:\n/serial/.../]
    subgraph modbus_exe["Your process"]
      subgraph serial["Library: remote_serial"]
        topic_serial --> driver["Serial port driver"]
      end
      code["Your code"] -- "DDS\n(potentially without\ncontext switch)" --> topic_serial
      code -- "or native API calls" ---> driver
      driver --> file{{"Character device"}}
    end
```


## Implementation details

The following diagram shows the high level view on the internals of this package:

```mermaid
flowchart TB
    owner["Native API"]
    external["ROS2 Interface API"]

    subgraph node["remote_serial::Node"]

      subgraph interface_ros2["remote_serial::Interface"]
        subgraph topics["Topics: /serial"]
          published_input[//inspect/input/]
          published_output[//inspect/output/]
        end

        subgraph services["Services: /serial"]
          service_send[//inject/output/]
          service_recv[//inject/input/]
        end
      end

      subgraph worker["remote_serial::Port"]
        output_queue["Output Queue"]
        thread["Worker thread\nrunning select()"]

        subgraph interface_native["remote_serial::Implementation"]
          subgraph register_input_cb["register_input_cb()"]
            input_cb["input_cb()"]
          end
          output("output()")
        end

        fd["File Descriptor"]
      end

    end
    file{{"Character device"}}

    %% owner -- "Constructor\nwith parameters" ----> node
    owner --> output --> output_queue
    %% owner --> register_input_cb

    external .-> published_input
    external .-> published_output
    external .-> service_send
    external .-> service_recv

    service_recv --> published_input
    service_recv --> input_cb

    service_send --> output_queue
    output_queue --> thread
    thread --> published_input
    thread --> published_output
    thread ---> input_cb --> owner

    file -- "input" ---> fd -- "input" ----> thread
    thread -- "output" --> fd -- "output" --> file

```

## Implementing serial line drivers

The serial line drivers implement the base class provided by this package.
For C++ drivers, that means that they use
`remote_serial::Implementation` as a parent class.
Its constructor requires an instance of `rclcpp::Node` to read the
parameters from.

<i>
Note:
There can be more than one instance of `remote_serial::Implementation`
per node. However they will all be controlled by the same parameters.
Though the default value of the parameter `serial_prefix` can be passed
into each individual constructor which is sufficient for common use cases.
</i>
<br/>
<br/>

The following methods need to be implemented by each driver:

```c++
public:
  // output writes bytes to the serial line
  virtual void output(const std::string &) override;

  // register_input_cb is used to set the callback function which
  // is called every time data is received from the serial line
  virtual void register_input_cb(void (*)(const std::string &msg,
                                          void *user_data),
                                 void *user_data) override;
```

The users can chose one of three ways to interact with child classes of `remote_serial::Implementation`:

- Link with the driver directly to make native API calls:

  ```c++
  auto serial = \
    std::make_shared<driver_package::DriverClass>(node, "/serial");
  serial->output("ATH+CHUP\r\n");
  ```

  In this case the driver is running within the same process and it is
  destroyed when the `serial` object is destroyed.

- Link with `remote_serial` to make ROS2 interface (DDS) calls
  (locally or over a network):

  ```c++
  auto serial = \
    std::make_shared<remote_serial::RemoteInterface>(node, "/serial");
  serial->output("ATH+CHUP\r\n");
  ```

  In this case the driver can be elsewhere within DDS' reach
  (same process or another side of the globe).

- Let the runtime make the choice between the above two options:

  ```c++
  auto serial = \
    remote_serial::Factory::New(node, "/serial");
  // or
  auto serial = \
    driver_package::Factory::New(node, "/serial");
  ```

  In this case the boolean value of the parameter `serial_is_remote`
  determines whether the driver is instantiated locally or if a remote
  interface is used to reach the driver instantiated elsewhere.
  Please, note, the trivial class `driver_package::Factory`
  (similar to `remote_serial::Factory`) has to be written
  to support this use case.

## Examples

### remote_microcontroller

See the UART implementation in
[remote_microcontroller](https://github.com/openvmp/microcontroller)
for an example of a driver that implements `remote_serial`.


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package remote_serial
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1.0.1 (2023-05-06)
------------------

1.0.0 (2023-05-01)
------------------
* Renamed the package to better align with REP-0144 guidelines
* Fixed missing class member initialization for the remote interface. Minor performance improvements.
* Compiles on foxy
* Refactored the boundary between the serial protocol implementation class and the serial port class as deemed necessary by an implementation of UART on a microcontroller.
* Flow control refactoring. Reduced overhead from debug logging.
* Refactored DDS path for all resources
* Work around a timing issue in the test cases
* Interface is refactored to support a factory and to coexist with other OpenVMP modules
* Made it work as a dependency for other packages. Made it work with binary data.
* updated the documentation to reflect the recent internal terminology change
* added the first test case and made it work
* First version that compiles
* added the openvmp banner
* added README.md
* Initial revision
* Contributors: Roman Kuzmenko


src/port.cpp
/*
 * OpenVMP, 2022
 *
 * Author: Roman Kuzmenko
 * Created: 2022-09-24
 *
 * Licensed under Apache License, Version 2.0.
 */

#define _DEFAULT_SOURCE
#include "remote_serial/port.hpp"

#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <string.h>
#include <sys/ioctl.h>
#include <termios.h>
#include <unistd.h>

#include "remote_serial/worker.hpp"

namespace remote_serial {

Port::Port(rclcpp::Node *node)
    : Implementation(node), port_settings_(new PortSettings()) {
  node_->declare_parameter("serial_dev_name", "/dev/ttyS0");
  node_->declare_parameter("serial_skip_init", false);
  node_->declare_parameter("serial_baud_rate", 115200);
  node_->declare_parameter("serial_data", 8);
  node_->declare_parameter("serial_parity", false);
  node_->declare_parameter("serial_stop", 1);
  node_->declare_parameter("serial_flow_control", false);
  node_->declare_parameter("serial_sw_flow_control", false);
  node_->declare_parameter("serial_bs", 1024);
  node_->get_parameter("serial_dev_name", port_settings_->dev_name);
  node_->get_parameter("serial_skip_init", port_settings_->skip_init);
  node_->get_parameter("serial_baud_rate", port_settings_->baud_rate);
  node_->get_parameter("serial_data", port_settings_->data);
  node_->get_parameter("serial_stop", port_settings_->stop);
  node_->get_parameter("serial_parity", port_settings_->parity);
  node_->get_parameter("serial_flow_control", port_settings_->flow_control);
  node_->get_parameter("serial_sw_flow_control",
                       port_settings_->sw_flow_control);
  node_->get_parameter("serial_bs", port_settings_->bs);

  RCLCPP_INFO(node_->get_logger(), "Serial node initialization complete for %s",
              port_settings_->dev_name.as_string().c_str());

  // Topics are initialized prior to the worker.
  // That is done to put the burden of null checks on the DDS side.
  // Should the burden be on the device driver side,
  // then the conditions for the serial line saturation will be met more often.
  worker_ = std::make_shared<Worker>(this, port_settings_);

  init_serial_();
}

void Port::output(const std::string &msg) { worker_->output(msg); }

// having pure pointers would improve performance here
// but it would be against the religion of so many
void Port::register_input_cb(void (*input_cb)(const std::string &msg,
                                              void *user_data),
                             void *user_data) {
  worker_->register_input_cb(input_cb, user_data);
}

void Port::inject_input(const std::string &msg) { worker_->inject_input(msg); }

int PortSettings::setup(int old_fd) {
  if (old_fd != -1) {
    close(old_fd);
  }

  int fd = ::open(dev_name.as_string().data(), O_RDWR);
  if (fd < 0) {
    throw std::invalid_argument("failed to open the device");
  }

  if (!skip_init.as_bool()) {
    // Get terminal attributes
    struct termios tty;
    ::memset(&tty, 0, sizeof(tty));
    if (::tcgetattr(fd, &tty) != 0) {
      ::close(fd);
      throw std::invalid_argument("failed to get the device attributes");
    }

    // Baud rate
    switch (baud_rate.as_int()) {
      case 0:
        cfsetispeed(&tty, B0);
        break;
      case 50:
        cfsetispeed(&tty, B50);
        break;
      case 75:
        cfsetispeed(&tty, B75);
        break;
      case 110:
        cfsetispeed(&tty, B110);
        break;
      case 134:
        cfsetispeed(&tty, B134);
        break;
      case 150:
        cfsetispeed(&tty, B150);
        break;
      case 200:
        cfsetispeed(&tty, B200);
        break;
      case 300:
        cfsetispeed(&tty, B300);
        break;
      case 600:
        cfsetispeed(&tty, B600);
        break;
      case 1200:
        cfsetispeed(&tty, B1200);
        break;
      case 1800:
        cfsetispeed(&tty, B1800);
        break;
      case 2400:
        cfsetispeed(&tty, B2400);
        break;
      case 4800:
        cfsetispeed(&tty, B4800);
        break;
      case 9600:
        cfsetispeed(&tty, B9600);
        break;
      case 19200:
        cfsetispeed(&tty, B19200);
        break;
      case 38400:
        cfsetispeed(&tty, B38400);
        break;
      case 57600:
        cfsetispeed(&tty, B57600);
        break;
      case 115200:
        cfsetispeed(&tty, B115200);
        break;
      case 230400:
        cfsetispeed(&tty, B230400);
        break;
      // case 460800:
      //   cfsetispeed(&tty, B460800);
      //   break;
      default:
        throw std::invalid_argument("unsupported baud rate");
    }

    // Data bits
    tty.c_cflag &= ~CSIZE;
    switch (data.as_int()) {
      case 5:
        tty.c_cflag |= CS5;
        break;
      case 6:
        tty.c_cflag |= CS6;
        break;
      case 7:
        tty.c_cflag |= CS7;
        break;
      case 8:
        tty.c_cflag |= CS8;
        break;
    }

    // Stop bits
    if (stop.as_int() == 1) {
      tty.c_cflag &= ~CSTOPB;
    } else {
      tty.c_cflag |= CSTOPB;
    }

    // Parity
    // TODO(clairbee): handle this as a dictionary to support odd parity
    if (parity.as_bool()) {
      tty.c_cflag |= PARENB;
      tty.c_cflag &= ~PARODD;
    } else {
      tty.c_cflag &= ~PARENB;
    }

    // Flow control
    if (flow_control.as_bool()) {
      tty.c_cflag |= CRTSCTS;  // Turn on h/w flow control
    } else {
      tty.c_cflag &= ~CRTSCTS;
    }
    if (sw_flow_control.as_bool()) {
      tty.c_iflag |= (IXON | IXOFF | IXANY);  // Turn on s/w flow ctrl
    } else {
      tty.c_iflag &= ~(IXON | IXOFF | IXANY);  // Turn off s/w flow ctrl
    }

    tty.c_cflag |= CREAD | CLOCAL;
    tty.c_iflag &= ~(IGNBRK | BRKINT | PARMRK | ISTRIP | INLCR | IGNCR | ICRNL);
    // Setting both to 0 will give a non-blocking read
    tty.c_cc[VTIME] = 0;
    tty.c_cc[VMIN] = 0;
    tty.c_lflag &= ~ICANON;  // Turn off canonical input
    tty.c_lflag &= ~(ECHO);  // Turn off echo
    tty.c_lflag &= ~ECHOE;   // Disable erasure
    tty.c_lflag &= ~ECHONL;  // Disable new-line echo
    tty.c_lflag &= ~ISIG;  // Disables recognition of INTR (interrupt), QUIT and
                           // SUSP (suspend) characters
    tty.c_oflag &= ~OPOST;  // Prevent special interpretation of output bytes
                            // (e.g. newline chars)
    tty.c_oflag &=
        ~ONLCR;  // Prevent conversion of newline to carriage return/line feed

    // Set the terminal attributes
    ::tcflush(fd, TCIFLUSH);
    if (::tcsetattr(fd, TCSANOW, &tty) != 0) {
      ::close(fd);
      throw std::invalid_argument("failed to set the device attributes");
    }
  }

  // Get the file descriptor attributes
#if defined(F_NOCACHE)
  ::fcntl(fd, F_NOCACHE, 1);
#endif
  int flags = ::fcntl(fd, F_GETFL, 0);
  if (flags == -1) {
    ::close(fd);
    throw std::invalid_argument("failed to get the file descriptor attributes");
  }

  flags |= O_NONBLOCK;

  // Set the file attributes
  if (::fcntl(fd, F_SETFL, flags)) {
    ::close(fd);
    throw std::invalid_argument("failed to set the file descriptor attributes");
  }

  return fd;
}

}  // namespace remote_serial
